extensions/wikia/SpecialMultipleLookup/MultiLookupPager.php

```diff
@@ -0,0 +1,163 @@
+<?php
+
+class MultiLookupPager extends TablePager {
+	const ML_TIMESTAMP = 'ml_ts';
+	const USERS = 'users';
+	const WIKI_URL = 'wiki_url';
+
+	/** @var string $target */
+	private $target;
+	/** @var MultiLookupRowFormatter $formatter */
+	private $formatter;
+
+	public function __construct( IContextSource $context, $target ) {
+		global $wgSpecialsDB;
+		parent::__construct( $context );
+
+		$this->target = IP::sanitizeIP( $target );
+		$this->mDb = wfGetDB( DB_SLAVE, [], $wgSpecialsDB );
+		$this->formatter = new MultiLookupRowFormatter( $context );
+	}
+
+	/**
+	 * This function should be overridden to provide all parameters
+	 * needed for the main paged query. It returns an associative
+	 * array with the following elements:
+	 *    tables => Table(s) for passing to Database::select()
+	 *    fields => Field(s) for passing to Database::select(), may be *
+	 *    conds => WHERE conditions
+	 *    options => option array
+	 *    join_conds => JOIN conditions
+	 *
+	 * @return array
+	 */
+	function getQueryInfo() {
+		return [
+			'tables' => [ 'multilookup' ],
+			'fields' => '*',
+			'conds' => [
+				'ml_ip_bin' => inet_pton( $this->target )
+			],
+		];
+	}
+
+	/**
+	 * Get per-wiki info for each row
+	 * @param array|object $row
+	 * @return string
+	 */
+	function formatRow( $row ) {
+		$wiki = WikiFactory::getWikiByID( $row->ml_city_id );
+
+		if ( $wiki ) {
+			$row->users = $this->getWikiUsers( $wiki->city_dbname );
+			$row->wiki_url = $wiki->city_url;
+		}
+
+		return parent::formatRow( $row );
+	}
+
+	/**
+	 * Get recent contributors that used this IP address on a wiki
+	 * @param string $dbName
+	 * @return string[] array of user names and IP addresses
+	 */
+	private function getWikiUsers( $dbName ) {
+		global $wgMemc;
+
+		$cacheKey = wfSharedMemcKey( 'multilookup', $this->target, $dbName );
+		$cachedValue = $wgMemc->get( $cacheKey );
+
+		if ( $cachedValue ) {
+			/** @var string[] $cachedValue */
+			return $cachedValue;
+		}
+
+		$dbr = wfGetDB( DB_SLAVE, [], $dbName );
+
+		$res = $dbr->select(
+			[ 'recentchanges' ],
+			[
+				'rc_user_text as user_name',
+				'rc_user as user_id',
+			],
+			[ 'rc_ip' => $this->target ],
+			__METHOD__,
+			[ 'DISTINCT' ]
+		);
+
+		$users = [];
+		foreach ( $res as $user ) {
+			// SUS-812: use username lookup - user ID for registered users, rc_user_text for anons
+			$userName = User::getUsername( $user->user_id, $user->user_name );
+
+			$users[] = $userName;
```

Comment:
Do we want this array to have unique values?

Reply:
`SELECT ... DISTINCT` should handle that part 🙂

Reply:
Yes, wanted to suggest the same :)